text/0088-contract-level-messages.md

```diff
@@ -0,0 +1,237 @@
+  * The number of topics a contract can register will be limited by a chainspec value.
+* emitting messages - `casper_emit_message`
+  * New messages are emitted under a pre-registered topic only.
+  * The message emitted will be of a pre-defined message type which is an enum that will initially support sending out human readable strings but can be extended in the future to support other types.
+
+When a new topic is registered through the `casper_manage_message_topic` FFI, the host creates a composite key derived from the callers entity address and the hash of the topic name.
+This step is required in order for users to be able to verify messages later on by querying global state using a `Key` derived as mentioned above.
+When querying global state for the topic control record, users can determine the number of messages that were emitted in the block that has the timestamp specified in the control record.
+
+When the host receives these messages through the `casper_emit_message` FFI, the identity of the caller entity will be attached to the message automatically by the host.
+The messages that are emitted will not be stored in global state; only a checksum of the message will be however stored. This allows users to verify the integrity of the message and also check that the message was indeed generated by the expected entity by doing a query to global state. The entries in global state are uniquely identified by the hash of the entity that emitted the message, the hash of the topic name and the index of the message.
+
+Messages will be passed by the execution engine to the node that will eventually send them out on the event stream after the execution is complete and committed.
+The messages sent out on the event stream contain the identity of the entity that emitted the message, the topic on which the message was emitted, the index of the message within the topic and the actual message payload.
+
+**Note:** the messages themselves cannot be reconstructed by reading the entry from global state since that entry contains only the checksum of the messages sent.
+
+## Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+The implementation for supporting contract level messages requires changes in both the execution engine and the node.
+
+### Changes to execution engine
+
+#### Introduce new FFIs:
+```rust
+/// Manages a message topic with the specified name for the calling contract.
+/// 
+/// The topic name is a string of a maximum length that is specified in the chainspec.
+/// The initially supported operation will be `Add` that registers a new message topic.
+/// Will return an error if:
+///     * the topic name length is exceeded
+///     * the topic already exists (in case of the Add operation)
+///     * the call was made from an `Account` context. We only allow stored contracts to register messages.
+pub fn casper_manage_message_topic(
```

Comment:
I'd remove this FFI in favour of registering all events at the contract installation step along side the entry points information. It will enable consumer to know what kind of event-data-schemas (at least names) they need before they start consuming events. Most of the time (all?) you know all the events your contract will emit at the compilation point. I'd even say that making on-the-fly topics is rather undesirable.

Reply:
Yes, we could do it but it would maybe reduce some of the flexibility. I'm guessing some contracts would rather delay registering topics until actually needed (because registering topics also cost gas). Also we would still want to have some kind of FFI for managing message topics in case we want to add some more operations in the future.

Reply:
> Why would registering topics on the fly be undesirable?

1. If that would be lazy, then emitting first event requires the topic registration and this additional gas. As someone who calls contract I want the gas costs to predictable and constant as much as possible.
2. If that would be lazy, then how do I know if the topic was already registered? You'd have to expose new `ffi` that checks if topic is already registered. Then before every call I'd have to make this check, so every event emit is two `ffi` calls.
3. Because of the above point nr 2, I'd have to register all the topics as soon as possible. Casper doesn't have constructors, but we very often mimics them by calling a `constructor` entrypoint just after the contract installation in `call` and then revoking the access to `constructor`, so it can't be called again. Not every contract needs calling a constructor (not calling it saves gas), but now I'd have to account for events registration and call the `constructor` also when contract needs topics registration.

Reply:
Changed the FFI to allow messages to be managed on contract install/upgrade as per request: https://github.com/casper-network/ceps/pull/88/commits/afaf653819ca6db9bf68415586ffd93aec2dd8a8